### DISCUSSION TITLE
How to change classes dynamically? (Tailwind with Next.js)

### DISCUSSION THREAD

**Discussion by danilo-leal** [tailwindlabs/tailwindcss#3461] [Help] [answered]
Hey guys!

I have a Next.js project setup with Tailwind and I've been trying to change some element's styling dynamically. I have a file at [my repo with some mock data](https://github.com/danilo-leal/coop.io/blob/main/components/demo/data.js). The data objects contain category and color values. They're displayed as such in the UI (the blue circle + text): 

<img width="312" alt="Screen Shot 2021-01-27 at 20 33 59" src="https://user-images.githubusercontent.com/67129314/106068470-7d3c2800-60df-11eb-8add-82b42eb9b925.png">

In the data file, I want to be able to set different colors to different categories. How do I do that with Tailwind classes? Furthermore, I'd like to render each component with the colors as declared in the data file. The component that consumes the data is the [**ExpensesEntry**](https://github.com/danilo-leal/coop.io/blob/main/components/demo/EarningsEntry.js).

I'm still learning React and JS in general so I'd appreciate some really slow instructions 😅

**Comment by mattatcubix** [comment]
I'm not using Next but I'd suggest using a ternary operator to conditionally change the CSS class is the best approach. eg.

`<p className={isBlueText ? 'text-blue-500' : 'text-black'}>`

**Reply by danilo-leal** [reply]
What if I had like 6 color options? How do I go about using this ternary route? Something like this?

`<p className={isBlueText ? 'text-blue-500 : 'text-black'}>`
`<p className={isGreenText ? 'text-green-500 : 'text-black'}>`
`<p className={isYellowText ? 'text-yellow-500 : 'text-black'}>`

etc?
It doesn't seem right tho 🤔

**Comment by alexanderbluhm** [comment]
@danilo-leal There are packages like [classnames](https://github.com/JedWatson/classnames) which are pretty useful for conditional classes.
Also make sure to use full class names (and no thing like `"text-" + isActive ? "green-500" : "red-500"`) in order not to violate purge css.

**Reply by aminBenSlimen** [reply]
what if i really need to use this way `"text-" + isActive ? "green-500" : "red-500"` ?

**Reply by picwellwisher12pk** [reply]
I want to make colors dynamic too. like `green-500` or `red-500` and then concatenate with background or text or border

**Reply by alexanderbluhm** [reply]
@picwellwisher12pk The preferred way is to do `isActive ? "text-green-500" : "text-red-500"` as you can also read [here](https://tailwindcss.com/docs/content-configuration#dynamic-class-names).

**Comment by simonswiss** [comment, marked as answer]
Hey @danilo-leal! 👋

Something I like to do is create a "lookup table" object with all the possible values of the `category` prop you would receive in that component:

```js
const textColor = {
  default: 'text-black',
  'Estrutural': 'text-green-500',
  'Investimento': 'text-blue-500',
  'Viagens': 'text-yellow-500',
  // ...
}
```

You can put this object somewhere at the top of your component. Then, in your markup, you can use `textColor[category]` to reach for the correct text color class name:

```html
<p className={textColor[category]}>I will receive the color set by the color prop lookup table!</p>
```

You can repeat the same process for the background color and anything you need to adapt to the category prop:

```js
const bgColor = {
  default: 'bg-black',
  'Estrutural': 'bg-green-500',
  'Investimento': 'bg-blue-500',
  'Viagens': 'bg-yellow-500',
  // ...
}
```

You could also have your lookup object return objects for each prop permutation:

```js
const colors = {
  default: {
    bg: 'bg-black',
    text: 'text-black'
  },
  'Estrutural': {
    bg:  'bg-green-500',
    text: 'text-green-500'
  },
  // ...
}
```

... and then reach for the nested value in the markup: 

```html
<div className={colors[category].bg}>
  <p className={colors[category].text}>I will be text of the category text color, on a div with the category bg color</p>
</div>
```

If you need to merge your "lookup" classes with other classes (very likely), you can use a package like `classnames`, or interpolate the color class with a template literal:

```html
<p className={`mt-4 text-2xl font-bold ${textColor[category]}`}>....</p>
```

All of this will still work with PurgeCSS, since the entire text colour class string is present in your file in the lookup table 👍

Hope it helps - I typed all this without really testing any of that. Hopefully there are no typos and it makes sense to you. 

I really enjoy this lookup table (not sure if that's the correct term 😅) approach, makes the code nice and clean and it's easy to go edit one f the classes for a category in the table object 👍

**Reply by danilo-leal** [reply]
Hey Simon! Dude, thanks a whole bunch for this write up, I appreciate it! I did exactly what you said and it didn't work out for some reason :( Here's the code I wrote:

```
const colors = {
    default: {
      bg: 'bg-gray-900',
      text: 'text-gray-900',
    },
    Estrutural: {
      bg: 'bg-blue-500',
      text: 'text-blue-500',
    },
    Investimento: {
      bg: 'bg-red-500',
      text: 'text-red-500',
    },
    Viagens: {
      bg: 'bg-green-500',
      text: 'text-green-500',
    },
  };

  return (
    <div className="flex flex-col space-y-4 pt-4">
      <div className="flex justify-between">
        <div className="flex space-x-3 items-center">
          <div className={`w-2 h-2 rounded-full ${colors[category.bg]}`}></div>
          <p className={`text-sm font-semibold ${colors[category.text]}`}>
            {category}
          </p>```

**Reply by simonswiss** [reply]
Hey!

You need to change the following in your code above (in return statement):

```diff
- ${colors[category.bg]}
+ ${colors[category].bg}
```

And the same for the text:

```diff
- ${colors[category.text]}
+ ${colors[category].text}
```

That _should_ work after that 👍

**Reply by danilo-leal** [reply]
It did work like a charm! Thanks a whole lot Simon :)

**Reply by elkarouani** [reply]
Declaring classes names with the cost of readability and filling the component with many class names or either thinking of isolating these names in a proper file, is not a pleasurable solution when we can use just pure defined CSS classes from our custom styles and what we use will be like this :  `${someState.category_name}-bg-color`; 
Or this : `${someState.category_name}-text-color`. 
Like what is better between filling with these class names in our components or in our custom CSS file.
And I will appreciate your suggestions and your opinions, regarding the readability of the source code and removing redundancy.

**Reply by a-y-u-s-h** [reply]
This solution only refactors out some classes in an object. It isn't truly 'dynamic' because I'm expecting to write `bg-${theme}-600` and expect theme variable to change and pick the color scheme described in my theme file. If it were truly dynamic, I would specify my colors within theme file only, I shouldn't have to make their className-mappings manually in an object outside of it.

**Reply by muhammederdinc** [reply]
@a-y-u-s-h Did you find a solution for the issue?

**Reply by a-y-u-s-h** [reply]
@muhammederdinc Nope. Apparently it's a purgeCSS thing. Purge should somehow know which classes are going to exist and it cannot do it if we put variables like `bg-${theme}-600`. I put all my props in an object these days, including `className` and then manipulate this object based on state changes, so my styles look like this now: 

```js
const theme = ...(just a string derived from some xstate statechart)

const styles = {
  container: {
    className: `
      w-full h-full overflow-hidden
      ${theme == "whatsapp" && "bg-whatsapp-500"}
      ${theme == "facebook" && "bg-facebook-500"}
      ${theme == "linear" && "bg-linear-500"}      
      ${theme == "jumanji" && "bg-jumanji-500"}       
    `
  }
}
```

Then in my JSX, I do this to pass in the props: `<div {...styles.container}></div>`

Another way could be to not use purge, but I figured that the above method gives me more flexibility so I just went with it without exploring more.

**Reply by muhammederdinc** [reply]
@a-y-u-s-h  I solved the problem as below.  But for a big project this solution didn't make much sense to me. But for now there seems to be nothing to be done (i am using Nuxt)
```
function backgroundColor(value) {
  return Object.values({
    ...(value === 'blue-silver-lake' && { class: 'bg-blue-silver-lake' }),
    ...(value === 'blue-yonder' && { class: 'bg-blue-silver-lake' }),
    ...(value === 'blue-little-boy' && { class: 'bg-blue-little-boy' }),
    ...(value === 'red-dark-coral' && { class: 'bg-red-dark-coral' }),
    ...(value === 'yellow-deep-saffron' && { class: 'bg-yellow-deep-saffron' })
  });
};
```

**Reply by hirwablessing** [reply]
> Hey @danilo-leal! 👋
> 
> Something I like to do is create a "lookup table" object with all the possible values of the `category` prop you would receive in that component:
> 
> ```js
> const textColor = {
>   default: 'text-black',
>   'Estrutural': 'text-green-500',
>   'Investimento': 'text-blue-500',
>   'Viagens': 'text-yellow-500',
>   // ...
> }
> ```
> 
> You can put this object somewhere at the top of your component. Then, in your markup, you can use `textColor[category]` to reach for the correct text color class name:
> 
> ```
> <p className={textColor[category]}>I will receive the color set by the color prop lookup table!</p>
> ```
> 
> You can repeat the same process for the background color and anything you need to adapt to the category prop:
> 
> ```js
> const bgColor = {
>   default: 'bg-black',
>   'Estrutural': 'bg-green-500',
>   'Investimento': 'bg-blue-500',
>   'Viagens': 'bg-yellow-500',
>   // ...
> }
> ```
> 
> You could also have your lookup object return objects for each prop permutation:
> 
> ```js
> const colors = {
>   default: {
>     bg: 'bg-black',
>     text: 'text-black'
>   },
>   'Estrutural': {
>     bg:  'bg-green-500',
>     text: 'text-green-500'
>   },
>   // ...
> }
> ```
> 
> ... and then reach for the nested value in the markup:
> 
> ```
> <div className={colors[category].bg}>
>   <p className={colors[category].text}>I will be text of the category text color, on a div with the category bg color</p>
> </div>
> ```
> 
> If you need to merge your "lookup" classes with other classes (very likely), you can use a package like `classnames`, or interpolate the color class with a template literal:
> 
> ```
> <p className={`mt-4 text-2xl font-bold ${textColor[category]}`}>....</p>
> ```
> 
> All of this will still work with PurgeCSS, since the entire text colour class string is present in your file in the lookup table 👍
> 
> Hope it helps - I typed all this without really testing any of that. Hopefully there are no typos and it makes sense to you.
> 
> I really enjoy this lookup table (not sure if that's the correct term 😅) approach, makes the code nice and clean and it's easy to go edit one f the classes for a category in the table object 👍

I did this but it doesn't work in Tailwind v3

**Reply by alexanderbluhm** [reply]
@hirwablessing This should also work in v3. Can you share your config and the code please?

**Comment by hirwablessing** [comment]
**Heading.tx**

```
import React, { ReactNode } from 'react'
import { colorStyle, fontSizeStyle, fontWeightStyle } from 'src/globals/global-values'
import { fontSize, fontWeight, Color } from 'src/types'

type HeadingProps = {
  fontSize?: fontSize
  fontWeight?: fontWeight
  color?: Color
  className?: string
  children: ReactNode
}

export default function Heading({
  fontSize = '5xl',
  fontWeight = 'bold',
  color = 'primary',
  className,
  children,
}: HeadingProps) {
  return (
    <h2
      className={`${fontSizeStyle[fontSize]}  ${fontWeightStyle[fontWeight]} ${
        colorStyle[color]
      } ${className && className} dark:text-main`}>
      {children}
    </h2>
  )
}
```

And then the class names object

```
export const fontWeightStyle: fontWeightStyleType = {
  normal: 'font-normal',
  regular: 'font-light',
  medium: 'font-medium',
  bold: 'font-bold',
}

export const bgStyle: bgStyleType = {
  primary: 'bg-primary',
  main: 'bg-main',
  error: 'bg-error-400',
  warning: 'bg-warning-400',
  success: 'bg-succes-400',
}

export const colorStyle: colorStyleType = {
  primary: 'text-primary',
  lightBlue: 'text-lightBlue',
  main: 'text-main',
  gray: 'text-gray-500',
  lightgray: 'text-gray-200',
}

export const fontSizeStyle: fontSizeStyleType = {
  xl: 'text-2xl',
  '2xl': 'text-2xl',
  '3xl': 'text-3xl',
  '4xl': 'text-4xl',
  '5xl': 'text-5xl',
}
```

Then this is my tailwind.config.js

```
module.exports = {
  content: ['./src/pages/**/*.{js,ts,jsx,tsx}', './src/components/**/*.{js,ts,jsx,tsx}'],
  darkMode: 'class',
  theme: {
    extend: {
      fontFamily: {
        primary: ['Inter', 'sans-serif'],
      },
      colors: {
        primary: '#060F33',
        lightBlue: '#454bba1a',
        main: '#FFFFFF',
      },
    },
  },
  variants: {
    extend: {},
  },
  plugins: [require('@tailwindcss/forms')],
}
```

**Reply by alexanderbluhm** [reply]
I also need the tailwind config file.
You have to ensure that the file where you export the styles is included in the `content` option, e.g. `content: ["./src/**/*.{html,js}"],`

**Reply by hirwablessing** [reply]
Thank you very much!
It worked now. I was missing putting the file where I was exporting the styles was missing in content. I added it like this:

```
  content: [
    './src/pages/**/*.{js,ts,jsx,tsx}',
    './src/components/**/*.{js,ts,jsx,tsx}',
    './src/globals/**/*.{js,ts,jsx,tsx}',
  ],
```

**Reply by hirwablessing** [reply]
But why do I need to put it in there? Does it mean that tailwind will watch for the changes?

**Reply by alexanderbluhm** [reply]
You can read about it here: https://tailwindcss.com/blog/just-in-time-the-next-generation-of-tailwind-css
(Or watch the video)

**Reply by hirwablessing** [reply]
I understand now. Thanks a lot, @alexanderbluhm :).

**Comment by jk-baseer** [comment]
Hey Guys, 
I read it it's answered. But I have further question exactly the same situation as the question with one small change. Wouldn't be nice if i create another thread ya? 

So in the above example, you have the lookup table to fill the values dynamically. Consider the product is substack similar one where

- The project is a single repo
- Every customisation (styling) for tenant is coming from server. Which means the CSS variables (example: H2 size) is not decided on build time but should be allowed on run time when the subdomain user want to change his H2 Size. 

Will this above approach still possible? 

Also read this from tailwind documentation: 
"Tailwind doesn’t include any sort of client-side runtime, so class names need to be statically extractable at build-time, and can’t depend on any sort of arbitrary dynamic values that change on the client. Use inline styles for these situations, or combine Tailwind with a CSS-in-JS library like Emotion if it makes sense for your project."
https://v2.tailwindcss.com/docs/just-in-time-mode#arbitrary-value-support 

Is this related to this scenario? 

Notes:
1. 
The easiest approach is applying the constants in root file.  but it's not scalable. 
`
let root = document.documentElement;
          const color = res.data.configs.app_color_primary;
          root.style.setProperty('--primary_color', color);
`



2.
second is trying to load a JSON file from Server and then take the values and apply in tailwind config. I think this is not totally possible. but we just tried and yes it says error like "typography undefined"
`
h1: {
          fontSize: theme(`fontSize.${style_data.typography.h1.size}`),
          fontWeight: theme(`fontWeight.${style_data.typography.boldText}`),
          letterSpacing: theme(
            `letterSpacing.${style_data.typography.h1.letterSpacing}`
          ),
          textDecoration: style_data.typography.decorationText,
        },
`

3. Trying the approach from this stackoverflow:  https://stackoverflow.com/a/63468001/1225899
Not able to fully get that method of using CSS variables, and root file setup. 

Looking for some helping hand on steps on what need to be done.

**Reply by hirwablessing** [reply]
@jk-baseer Is it possible to define all the possible H2 Sizes? Or they're unpredictable?

**Reply by jk-baseer** [reply]
Yes we can do that. Here is the User story. 
- A user should be able to select the size from his profile builder
- Developer should be able to load this H2 style based on the user selection in browser (UNDER SUB DOMAIN). It's a single code base

**Reply by hirwablessing** [reply]
If you have all the available H2 Sizes that the user can choose from then you should probably define an object that contains them as keys and then put values on each key so that you might use them dynamically in your classes. Did you check my example above? Let me know if it's a different case for you.

**Comment by MichaelAllenWarner** [comment]
Here is a TypeScript twist on @simonswiss's lookup-table solution:

```typescript
// some central file that stores all color options
export enum Color {
  Green = 'green-500',
  Blue = 'blue-500',
  // ...
}
```

```typescript
// some other central file that components can use for bg-color options
import { Color } from /* ... */

export const bgColor: { [C in Color]: `bg-${C}` } = {
  'green-500': 'bg-green-500',
  'blue-500': 'bg-blue-500',
  // ...
};
```

Now if you ever change `Color`, TypeScript will force you to also change `bgColor` accordingly. Oh, and you get the "Add missing properties" quick-fix option on `bgColor` (in VS Code, anyway), so you can update it without even typing out the key-value pairs.

(You can use a union-type instead of an enum if you prefer.)

Final touch: fiddle with [settings](https://github.com/tailwindlabs/tailwindcss/issues/7553#issuecomment-735915659) to get Tailwind IntelliSense on those classes.
